File: lib/property_rights/domain/jurisdiction_policy.dart
enum NorthAmericaCountry {
  canada('CA'),
  unitedStates('US'),
  mexico('MX');

  const NorthAmericaCountry(this.code);

  final String code;

  static NorthAmericaCountry fromCode(String value) {
    final normalized = value.trim().toUpperCase();
    return NorthAmericaCountry.values.firstWhere(
      (country) => country.code == normalized,
      orElse: () => throw FormatException(
        'Unsupported North American country code: $value',
      ),
    );
  }
}

enum JurisdictionPolicyStatus {
  designOnly,
  pendingApproval,
  active,
  suspended,
  retired,
}

enum ControlledPropertyFeature {
  propertyDraftIntake,
  publicPropertyListings,
  propertyOffers,
  rightsListings,
  regulatedEnergyAssets,
  businessSales,
  clientFunds,
}

class JurisdictionKey {
  JurisdictionKey({
    required this.country,
    String? subdivisionCode,
  }) : subdivisionCode = _normalizeSubdivision(subdivisionCode);

  final NorthAmericaCountry country;
  final String? subdivisionCode;

  bool get isCountryBaseline => subdivisionCode == null;

  String get value => subdivisionCode == null
      ? country.code
      : '${country.code}-$subdivisionCode';

  Map<String, Object?> toMap() => <String, Object?>{
        'countryCode': country.code,
        'subdivisionCode': subdivisionCode,
      };

  factory JurisdictionKey.fromMap(Map<String, Object?> map) {
    return JurisdictionKey(
      country: NorthAmericaCountry.fromCode(map['countryCode'] as String),
      subdivisionCode: map['subdivisionCode'] as String?,
    );
  }

  static String? _normalizeSubdivision(String? value) {
    final normalized = value?.trim().toUpperCase();
    return normalized == null || normalized.isEmpty ? null : normalized;
  }

  @override
  bool operator ==(Object other) =>
      other is JurisdictionKey &&
      other.country == country &&
      other.subdivisionCode == subdivisionCode;

  @override
  int get hashCode => Object.hash(country, subdivisionCode);

  @override
  String toString() => value;
}

class ResponsibleBrokerageEntity {
  const ResponsibleBrokerageEntity({
    required this.id,
    required this.legalName,
    required this.country,
    required this.active,
  });

  final String id;
  final String legalName;
  final NorthAmericaCountry country;
  final bool active;

  Map<String, Object?> toMap() => <String, Object?>{
        'id': id,
        'legalName': legalName,
        'countryCode': country.code,
        'active': active,
      };

  factory ResponsibleBrokerageEntity.fromMap(Map<String, Object?> map) {
    return ResponsibleBrokerageEntity(
      id: map['id'] as String,
      legalName: map['legalName'] as String,
      country: NorthAmericaCountry.fromCode(map['countryCode'] as String),
      active: map['active'] as bool? ?? false,
    );
  }
}

class BrokerageLicense {
  BrokerageLicense({
    required this.id,
    required this.entityId,
    required this.jurisdiction,
    required this.licenseNumber,
    required this.validFrom,
    this.validUntil,
    required this.active,
  });

  final String id;
  final String entityId;
  final JurisdictionKey jurisdiction;
  final String licenseNumber;
  final DateTime validFrom;
  final DateTime? validUntil;
  final bool active;

  bool isValidAt(DateTime at) {
    return active &&
        !at.isBefore(validFrom) &&
        (validUntil == null || at.isBefore(validUntil!));
  }

  Map<String, Object?> toMap() => <String, Object?>{
        'id': id,
        'entityId': entityId,
        'jurisdiction': jurisdiction.toMap(),
        'licenseNumber': licenseNumber,
        'validFrom': validFrom.toUtc().toIso8601String(),
        'validUntil': validUntil?.toUtc().toIso8601String(),
        'active': active,
      };

  factory BrokerageLicense.fromMap(Map<String, Object?> map) {
    return BrokerageLicense(
      id: map['id'] as String,
      entityId: map['entityId'] as String,
      jurisdiction: JurisdictionKey.fromMap(
        Map<String, Object?>.from(map['jurisdiction'] as Map),
      ),
      licenseNumber: map['licenseNumber'] as String,
      validFrom: DateTime.parse(map['validFrom'] as String),
      validUntil: map['validUntil'] == null
          ? null
          : DateTime.parse(map['validUntil'] as String),
      active: map['active'] as bool? ?? false,
    );
  }
}

class JurisdictionFeaturePolicy {
  JurisdictionFeaturePolicy({
    Set<ControlledPropertyFeature> enabledFeatures = const {},
  }) : enabledFeatures = Set.unmodifiable(enabledFeatures);

  final Set<ControlledPropertyFeature> enabledFeatures;

  bool enables(ControlledPropertyFeature feature) =>
      enabledFeatures.contains(feature);

  Map<String, Object?> toMap() => <String, Object?>{
        'enabledFeatures': enabledFeatures
            .map((feature) => feature.name)
            .toList(growable: false)
          ..sort(),
      };

  factory JurisdictionFeaturePolicy.fromMap(Map<String, Object?> map) {
    final raw = (map['enabledFeatures'] as List<Object?>? ?? const [])
        .whereType<String>();
    return JurisdictionFeaturePolicy(
      enabledFeatures: raw
          .map(
            (name) => ControlledPropertyFeature.values.firstWhere(
              (feature) => feature.name == name,
              orElse: () => throw FormatException(
                'Unsupported controlled property feature: $name',
              ),
            ),
          )
          .toSet(),
    );
  }
}

class FeatureDecision {
  FeatureDecision({
    required this.allowed,
    Iterable<String> reasons = const [],
  }) : reasons = List.unmodifiable(reasons);

  final bool allowed;
  final List<String> reasons;
}

class JurisdictionPolicy {
  JurisdictionPolicy({
    required this.schemaVersion,
    required this.id,
    required this.jurisdiction,
    required this.status,
    required this.responsibleEntity,
    required this.features,
    this.brokerageLicense,
    this.complianceOwnerId,
    this.requiredFormSetVersion,
    this.legalReviewVersion,
    this.trustFundsApprovalVersion,
    this.effectiveAt,
    this.expiresAt,
  });

  final int schemaVersion;
  final String id;
  final JurisdictionKey jurisdiction;
  final JurisdictionPolicyStatus status;
  final ResponsibleBrokerageEntity responsibleEntity;
  final JurisdictionFeaturePolicy features;
  final BrokerageLicense? brokerageLicense;
  final String? complianceOwnerId;
  final String? requiredFormSetVersion;
  final String? legalReviewVersion;
  final String? trustFundsApprovalVersion;
  final DateTime? effectiveAt;
  final DateTime? expiresAt;

  List<String> validationIssues(DateTime at) {
    final issues = <String>[];
    if (schemaVersion != 1) {
      issues.add('Unsupported jurisdiction policy schema version.');
    }
    if (id.trim().isEmpty) {
      issues.add('Policy identifier is missing.');
    }
    if (jurisdiction.isCountryBaseline) {
      issues.add(
        'Country baseline policies cannot authorize production features.',
      );
    }
    if (status != JurisdictionPolicyStatus.active) {
      issues.add('Jurisdiction policy is not active.');
    }
    if (!responsibleEntity.active) {
      issues.add('Responsible brokerage entity is not active.');
    }
    if (responsibleEntity.country != jurisdiction.country) {
      issues.add('Responsible entity does not match the jurisdiction country.');
    }
    if (brokerageLicense == null) {
      issues.add('Brokerage licence is missing.');
    } else {
      if (brokerageLicense!.entityId != responsibleEntity.id) {
        issues.add('Brokerage licence belongs to a different entity.');
      }
      if (brokerageLicense!.jurisdiction != jurisdiction) {
        issues.add('Brokerage licence does not match the jurisdiction.');
      }
      if (!brokerageLicense!.isValidAt(at)) {
        issues.add('Brokerage licence is not valid at the requested time.');
      }
    }
    if ((complianceOwnerId ?? '').trim().isEmpty) {
      issues.add('Supervising compliance owner is missing.');
    }
    if ((requiredFormSetVersion ?? '').trim().isEmpty) {
      issues.add('Approved form set version is missing.');
    }
    if ((legalReviewVersion ?? '').trim().isEmpty) {
      issues.add('Legal review version is missing.');
    }
    if (effectiveAt == null || at.isBefore(effectiveAt!)) {
      issues.add('Policy is not yet effective.');
    }
    if (expiresAt != null && !at.isBefore(expiresAt!)) {
      issues.add('Policy has expired.');
    }
    if (features.enables(ControlledPropertyFeature.clientFunds) &&
        (trustFundsApprovalVersion ?? '').trim().isEmpty) {
      issues.add(
        'Client-funds handling requires a separate trust approval version.',
      );
    }
    return issues;
  }

  FeatureDecision decisionFor(
    ControlledPropertyFeature feature, {
    required DateTime at,
  }) {
    final reasons = validationIssues(at);
    if (!features.enables(feature)) {
      reasons.add('Feature ${feature.name} is disabled for this jurisdiction.');
    }
    return FeatureDecision(
      allowed: reasons.isEmpty,
      reasons: reasons,
    );
  }

  Map<String, Object?> toMap() => <String, Object?>{
        'schemaVersion': schemaVersion,
        'id': id,
        'jurisdiction': jurisdiction.toMap(),
        'status': status.name,
        'responsibleEntity': responsibleEntity.toMap(),
        'features': features.toMap(),
        'brokerageLicense': brokerageLicense?.toMap(),
        'complianceOwnerId': complianceOwnerId,
        'requiredFormSetVersion': requiredFormSetVersion,
        'legalReviewVersion': legalReviewVersion,
        'trustFundsApprovalVersion': trustFundsApprovalVersion,
        'effectiveAt': effectiveAt?.toUtc().toIso8601String(),
        'expiresAt': expiresAt?.toUtc().toIso8601String(),
      };

  factory JurisdictionPolicy.fromMap(Map<String, Object?> map) {
    final licenseMap = map['brokerageLicense'];
    return JurisdictionPolicy(
      schemaVersion: map['schemaVersion'] as int,
      id: map['id'] as String,
      jurisdiction: JurisdictionKey.fromMap(
        Map<String, Object?>.from(map['jurisdiction'] as Map),
      ),
      status: JurisdictionPolicyStatus.values.firstWhere(
        (status) => status.name == map['status'],
        orElse: () => throw FormatException(
          'Unsupported jurisdiction policy status: ${map['status']}',
        ),
      ),
      responsibleEntity: ResponsibleBrokerageEntity.fromMap(
        Map<String, Object?>.from(map['responsibleEntity'] as Map),
      ),
      features: JurisdictionFeaturePolicy.fromMap(
        Map<String, Object?>.from(map['features'] as Map),
      ),
      brokerageLicense: licenseMap == null
          ? null
          : BrokerageLicense.fromMap(
              Map<String, Object?>.from(licenseMap as Map),
            ),
      complianceOwnerId: map['complianceOwnerId'] as String?,
      requiredFormSetVersion: map['requiredFormSetVersion'] as String?,
      legalReviewVersion: map['legalReviewVersion'] as String?,
      trustFundsApprovalVersion: map['trustFundsApprovalVersion'] as String?,
      effectiveAt: map['effectiveAt'] == null
          ? null
          : DateTime.parse(map['effectiveAt'] as String),
      expiresAt: map['expiresAt'] == null
          ? null
          : DateTime.parse(map['expiresAt'] as String),
    );
  }
}

File: firebase/functions/jurisdiction_policy.js
"use strict";

const COUNTRY_CODES = new Set(["CA", "US", "MX"]);
const POLICY_STATUSES = new Set([
  "designOnly",
  "pendingApproval",
  "active",
  "suspended",
  "retired",
]);
const FEATURES = new Set([
  "propertyDraftIntake",
  "publicPropertyListings",
  "propertyOffers",
  "rightsListings",
  "regulatedEnergyAssets",
  "businessSales",
  "clientFunds",
]);

function asDate(value) {
  if (!value) return null;
  const parsed = value instanceof Date ? value : new Date(value);
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

function jurisdictionKey(jurisdiction) {
  if (!jurisdiction || !COUNTRY_CODES.has(jurisdiction.countryCode)) {
    return null;
  }
  const subdivision = String(jurisdiction.subdivisionCode || "")
      .trim()
      .toUpperCase();
  return subdivision ?
    `${jurisdiction.countryCode}-${subdivision}` :
    jurisdiction.countryCode;
}

function validatePolicy(policy, at = new Date()) {
  const issues = [];
  const key = jurisdictionKey(policy && policy.jurisdiction);
  const entity = policy && policy.responsibleEntity;
  const license = policy && policy.brokerageLicense;
  const enabled = new Set(
      policy && policy.features && policy.features.enabledFeatures || [],
  );

  if (!policy || policy.schemaVersion !== 1) {
    issues.push("Unsupported jurisdiction policy schema version.");
  }
  if (!policy || !String(policy.id || "").trim()) {
    issues.push("Policy identifier is missing.");
  }
  if (!key) {
    issues.push("Jurisdiction is invalid.");
  } else if (!key.includes("-")) {
    issues.push(
        "Country baseline policies cannot authorize production features.",
    );
  }
  if (!policy || !POLICY_STATUSES.has(policy.status)) {
    issues.push("Jurisdiction policy status is invalid.");
  } else if (policy.status !== "active") {
    issues.push("Jurisdiction policy is not active.");
  }
  if (!entity || entity.active !== true) {
    issues.push("Responsible brokerage entity is not active.");
  } else if (
    policy.jurisdiction &&
    entity.countryCode !== policy.jurisdiction.countryCode
  ) {
    issues.push("Responsible entity does not match the jurisdiction country.");
  }
  if (!license) {
    issues.push("Brokerage licence is missing.");
  } else {
    if (!entity || license.entityId !== entity.id) {
      issues.push("Brokerage licence belongs to a different entity.");
    }
    if (jurisdictionKey(license.jurisdiction) !== key) {
      issues.push("Brokerage licence does not match the jurisdiction.");
    }
    const validFrom = asDate(license.validFrom);
    const validUntil = asDate(license.validUntil);
    if (
      license.active !== true ||
      !validFrom ||
      at < validFrom ||
      (validUntil && at >= validUntil)
    ) {
      issues.push("Brokerage licence is not valid at the requested time.");
    }
  }
  if (!String(policy && policy.complianceOwnerId || "").trim()) {
    issues.push("Supervising compliance owner is missing.");
  }
  if (!String(policy && policy.requiredFormSetVersion || "").trim()) {
    issues.push("Approved form set version is missing.");
  }
  if (!String(policy && policy.legalReviewVersion || "").trim()) {
    issues.push("Legal review version is missing.");
  }
  const effectiveAt = asDate(policy && policy.effectiveAt);
  const expiresAt = asDate(policy && policy.expiresAt);
  if (!effectiveAt || at < effectiveAt) {
    issues.push("Policy is not yet effective.");
  }
  if (expiresAt && at >= expiresAt) {
    issues.push("Policy has expired.");
  }
  for (const feature of enabled) {
    if (!FEATURES.has(feature)) {
      issues.push(`Unsupported controlled property feature: ${feature}.`);
    }
  }
  if (
    enabled.has("clientFunds") &&
    !String(policy && policy.trustFundsApprovalVersion || "").trim()
  ) {
    issues.push(
        "Client-funds handling requires a separate trust approval version.",
    );
  }
  return issues;
}

function canUseFeature(policy, feature, at = new Date()) {
  const reasons = validatePolicy(policy, at);
  const enabled = new Set(
      policy && policy.features && policy.features.enabledFeatures || [],
  );
  if (!FEATURES.has(feature)) {
    reasons.push(`Unknown controlled property feature: ${feature}.`);
  } else if (!enabled.has(feature)) {
    reasons.push(`Feature ${feature} is disabled for this jurisdiction.`);
  }
  return {
    allowed: reasons.length === 0,
    reasons,
  };
}

function resolveExactPolicy(policies, jurisdiction) {
  const requestedKey = jurisdictionKey(jurisdiction);
  if (!requestedKey) return null;
  return policies.find(
      (policy) => jurisdictionKey(policy.jurisdiction) === requestedKey,
  ) || null;
}

module.exports = {
  COUNTRY_CODES,
  FEATURES,
  POLICY_STATUSES,
  canUseFeature,
  jurisdictionKey,
  resolveExactPolicy,
  validatePolicy,
};

